desc/coils.py

```diff
@@ -430,6 +485,185 @@ def from_symmetry(cls, coils, NFP, sym=False):
+        coilinds = [2]  # always start at the 3rd line
+        names = []
+
+        # read in the coils file
+        with open(coil_file) as f:
+            lines = f.readlines()
+            for i, line in enumerate(lines):
+                if (
+                    line.find("periods") != -1
+                    or line.find("begin filament") != -1
+                    or line.find("end") != -1
+                    or line.find("mirror") != -1
+                ):
+                    continue  # skip headers and last line
+                if (
+                    len(line.split()) != 4
+                    and line.strip()  # ensure not counting blank lines
+                ):
+                    coilinds.append(i)
+                    names.append(" ".join(line.split()[4:]))
+        if len(lines[3].split()) != 4:
+            raise OSError(
+                "4th line in file must be the start of the first coil!"
+                + "Expected a line of length 4 (after .split()),"
+                + f"instead got length {lines[3].split()}"
+            )
+        header_lines_not_as_expected = np.array(
+            [
+                len(lines[0].split()) != 2,
+                len(lines[1].split()) != 2,
+                len(lines[2].split()) != 2,
+            ]
+        )
+        if np.any(header_lines_not_as_expected):
+            raise OSError(
+                "first 3 lines in file must be the header lines,"
+                + " each of length 2 (after .split())!"
+                + f"Line(s) {lines[np.where(header_lines_not_as_expected)[0]]}"
+                + " is not length 2"
+            )
+
+        for i, (start, end) in enumerate(zip(coilinds[0:-1], coilinds[1:])):
+            coords = np.genfromtxt(lines[start + 1 : end])
+            if i % 20 == 0:
+                print("reading coil " + f"{i}")
+
+            tempx = np.append(coords[:, 0], np.array([coords[0, 0]]))
+            tempy = np.append(coords[:, 1], np.array([coords[0, 1]]))
+            tempz = np.append(coords[:, 2], np.array([coords[0, 2]]))
+
+            coils.append(
+                XYZCoil(
+                    coords[:, -1][0],
+                    tempx,
+                    tempy,
+                    tempz,
+                    grid=grid,
+                    method=method,
+                    name=names[i],
+                )
+            )
+
+        return CoilSet(*coils)
+
+    def save_in_MAKEGRID_format(self, coilsFilename, NFP=1, grid=None):
+        """Save CoilSet of as a MAKEGRID-formatted coil txtfile.
+
+        By default, each coil is assigned to the same Coilgroup in MAKEGRID
+        with the name "Modular". For more details see the MAKEGRID documentation
+        https://princetonuniversity.github.io/STELLOPT/MAKEGRID.html
+
+        Parameters
+        ----------
+        filename : str or path-like
+            path save CoilSet as a file in MAKEGRID txt format
+        NFP : int, default 1
```

Comment:
this seems a bit confusing to the way we normally use NFP. Are we always assuming that the coilset goes around the whole plasma? or does the coilset only cover 1 field period?

Reply:
by assumption yea this is assuming it goes around the whole plasma, if NFP=1 (think of it as NFP of the coilset).

 If NFP is larger than 1, then only the first len(coils)/NFP are unique, and so it saves only those ones (though I just realized I need to change the from_makegrid method to work correctly with that

Reply:
OK for now I guess. Still feels a bit confusing. I'm wondering if maybe a coilset by definition should only have coils over 1 period, but have an NFP attribute saying how many times it repeats, and when you call `compute_magnetic_field` it automatically sums over the different periods. That also eliminates the need to have explicit constraints that coil 1 is the same thing as coil 1+NFP, just rotated.

Reply:
I guess that would only work for modular coils though...

Reply:
It is confusing I agree and I dont like how I've defined it. I like that better, I think I'll probably change to that

Reply:
your way could work for non modular coils if we just make sure the NFP is one for them...